File: src/main/java/com/mikea/gae/rx/RxOfy.java
package com.mikea.gae.rx;

import com.google.common.base.Function;
import com.googlecode.objectify.Key;
import com.mikea.gae.rx.base.IAction;

import javax.annotation.Nullable;

import static com.googlecode.objectify.ObjectifyService.ofy;

/**
 * @author mike.aizatsky@gmail.com
 */
public class RxOfy {
    private RxOfy() { }

    public static <T> IAction<T> save() {
        return new IAction<T>() {
            @Override
            public void perform(T t) {
                ofy().save().entity(t);
            }
        };
    }

    public static <T> Function<Key<T>, T> loadSafe() {
        return new Function<Key<T>, T>() {
            @Nullable
            @Override
            public T apply(@Nullable Key<T> input) {
                return ofy().load().key(input).safe();
            }
        };
    }
}

File: src/main/scala/com/mikea/gae/rx/RxOfy.scala
package com.mikea.gae.rx

import com.googlecode.objectify.Key
import com.googlecode.objectify.Result
import com.googlecode.objectify.ObjectifyService.ofy

/**
 * @author mike.aizatsky@gmail.com
 */
object RxOfy {
  def save[T]: (T) => Result[Key[T]] = (t: T) => ofy.save.entity(t)

  def loadSafe[T]: (Key[T]) => T = (key : Key[T]) => ofy.load.key(key).safe

  def saveMulti[T]: (Iterable[T]) => Map[Key[T], T] = (values : Iterable[T]) => {
    import scala.collection.JavaConverters._
    ofy.save.entities(values.asJava).now().asScala.toMap
  }
}